## ques 2 a calculate_P_n is basically calculating the probability that no one share their birthday , 1- pn gives the probability of atleast two people sharing birthday. 

#### Probability of atleast 2 people sharing birthday is =  1 -  P( all people have unique birth dates)
#### P = 1 - (365/365 *( 365-1/365) *( (365 -1 - 1)/365) *(...) *(365-i-1/365)) 
#### This will give the value from the function calculate_P_n(n): where n = i

In [15]:
def calculate_P_n(n):
    P_n = 1
    for i in range(n):
        P_n *= (365 - i) / 365
    return P_n

n = 23
result =1- calculate_P_n(n)
print(f"P_{n} is approximately {result:.10f}")


P_23 is approximately 0.5072972343


P_22 is approximately 0.4757
P_23 is approximately 0.5073

-------------------------------

Ques 2) b) basically iterate for all values of number of value until the probability  at least two will share a birthday is more than 75% then break

In [16]:
n = 1
while(True):
  result =1- calculate_P_n(n)
  if( result > 0.75) :
    print("minimum value of n such that the probability is just more than 75% is", n)
    break
  else:
    n += 1
  
  

minimum value of n such that the probability is just more than 75% is 32


### Similary here we are given the value of probability and we have to find least n value for the same
#### It can be done using starting from n = 1 till where the probability is just more than 75% we break the loop

# Answer is 32

---------------------

ques 2 c ) 

## Here, its given that a person's fake date is generated from the given real date of the user so we have 365-1 dates from which a fake date is chosen and to make sure it's unique all the date which were taken before  

- So considering we have 2 humans initially  , these are written in such a way to make sure that all of thee chosen dates are unique
- P1Realdate = (365/365)
- P1FakeDate = (365-1)/364
- P2RealDate = ( 365-2)/365
- P2FakeDate = ( 365 - 3)/364
- All the unique probabilities are to be multiplied to get the probability that no 
- = (P1Realdata) * (P1FakeData) * (P2RealData) * (P2FakeDate)

- The probabiltuy that all of these have atleast 2 of them matching is 
- 1 - (P1Realdata) * (P1FakeData) * (P2RealData) * (P2FakeDate)

- This should be 
- 1 - (P1Realdata) * (P1FakeData) * (P2RealData) * (P2FakeDate) >= p
- 1 - p >=  (P1Realdata) * (P1FakeData) * (P2RealData) * (P2FakeDate)
- So the loop is made to break for the above condition 
and works till 1- p < (P1Realdata) * (P1FakeData) * (P2RealData) * (P2FakeDate) for a given probabilty

In [2]:
import math as mth
def compute_birthday_paradox_probability(p):
    n = 1
    probability_no_match = 1
    remaining_unique_dates = 365
    while probability_no_match > (1 - p):
        probability_no_match *= (remaining_unique_dates ) /365
        remaining_unique_dates -= 1
        probability_no_match *= (remaining_unique_dates )/364
        remaining_unique_dates -= 1  
        n += 1

    return n

p = 0.49
result = compute_birthday_paradox_probability(p)
print(f"The smallest integer 'n' for probability {p} is {result}.")


The smallest integer 'n' for probability 0.49 is 13.


-----------------------------

In [18]:
print(1.645*1.645)
print("b = ", 220+2.706025 )
print("c = ",110*110 )

2.706025
b =  222.706025
c =  12100


ans for second part n = 32

In [19]:
import numpy as np

def quadratic_inequality_solver(a, b, c):

  # Calculate the discriminant
  discriminant = b**2 - 4 * a * c

  # If the discriminant is positive, the roots are real and distinct
  if discriminant > 0:
    root1 = (-b + np.sqrt(discriminant)) / (2 * a)
    root2 = (-b - np.sqrt(discriminant)) / (2 * a)
    return [root1, root2]

  # If the discriminant is zero, the roots are real and equal
  elif discriminant == 0:
    root = -b / (2 * a)
    return [root, root]

  # If the discriminant is negative, the roots are complex
  else:
    return ["complex", "complex"]

# Example usage
a = 1
b = 1.645
c = -110

roots = quadratic_inequality_solver(a, b, c)

print("The roots of the quadratic inequality are:", roots)


The roots of the quadratic inequality are: [9.697790217004473, -11.342790217004472]


In [20]:
import numpy as np

def quadratic_inequality_solver(a, b, c):

  # Calculate the discriminant
  discriminant = b**2 - 4 * a * c

  # If the discriminant is positive, the roots are real and distinct
  if discriminant > 0:
    root1 = (-b + np.sqrt(discriminant)) / (2 * a)
    root2 = (-b - np.sqrt(discriminant)) / (2 * a)
    return [root1, root2]

  # If the discriminant is zero, the roots are real and equal
  elif discriminant == 0:
    root = -b / (2 * a)
    return [root, root]

  # If the discriminant is negative, the roots are complex
  else:
    return ["complex", "complex"]

# Example usage
a = 1
b = 1.645
c = -110

roots = quadratic_inequality_solver(a, b, c)

print("The roots of the quadratic inequality are:", roots)


The roots of the quadratic inequality are: [9.697790217004473, -11.342790217004472]


√(n) =  9.697790217004473 is only possible value so 

In [21]:
9.697790217004473*9.697790217004473

94.04713509302765